File: mobile_test_backup_auth/mobile/lib/api_client.dart
import 'dart:convert';
import 'dart:io';

import 'package:crypto/crypto.dart';

class BuddiesApiClient {
  BuddiesApiClient({
    String baseUrl = const String.fromEnvironment(
      'API_BASE_URL',
      defaultValue: 'http://10.0.2.2:8080/api',
    ),
    HttpClient? httpClient,
  })  : _baseUri = Uri.parse(baseUrl),
        _httpClient = httpClient ?? HttpClient() {
    _assertSecureBaseUri(_baseUri);
  }

  final Uri _baseUri;
  final HttpClient _httpClient;

  Future<List<LobbySummary>> getLobbies({String? accessToken}) async {
    final body = await _requestJson('/lobbies', accessToken: accessToken);
    return (body as List<dynamic>)
        .map((item) => LobbySummary.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<LoginResponse> login(String email, String password) async {
    // Passwords are sent only over HTTPS/TLS and must be bcrypt-verified by the server.
    final body = await _requestJson(
      '/auth/login',
      method: 'POST',
      body: {'email': email, 'password': password},
    );
    return LoginResponse.fromJson(body as Map<String, dynamic>);
  }

  Future<void> verifySignup(String email, String otp) async {
    await _requestJson(
      '/auth/signup/verify',
      method: 'POST',
      body: {'email': email, 'otp': sha256Hex(otp)},
    );
  }

  Future<void> confirmPasswordReset(String token, String newPassword, String newPasswordConfirm) async {
    await _requestJson(
      '/auth/password-reset/confirm',
      method: 'POST',
      body: {
        'token': sha256Hex(token),
        'newPassword': newPassword,
        'newPasswordConfirm': newPasswordConfirm,
      },
    );
  }

  Future<dynamic> _requestJson(
    String path, {
    String method = 'GET',
    String? accessToken,
    Map<String, dynamic>? body,
  }) async {
    final request = await _openRequest(method, _baseUri.resolve(_joinPath(path)));
    request.headers.set(HttpHeaders.acceptHeader, 'application/json');
    request.headers.set(HttpHeaders.contentTypeHeader, 'application/json');
    request.headers.set(HttpHeaders.cacheControlHeader, 'no-store');
    request.headers.set(HttpHeaders.pragmaHeader, 'no-cache');

    // TODO: Store JWTs in platform secure storage after the auth flow is implemented.
    if (accessToken != null) {
      request.headers.set(HttpHeaders.authorizationHeader, 'Bearer $accessToken');
    }
    if (body != null) {
      request.write(jsonEncode(body));
    }

    final response = await request.close();
    final text = await response.transform(utf8.decoder).join();
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw ApiException(response.statusCode, text);
    }
    if (text.isEmpty) {
      return null;
    }
    return jsonDecode(text);
  }

  String _joinPath(String path) {
    final normalizedBase = _baseUri.path.endsWith('/') ? _baseUri.path : '${_baseUri.path}/';
    final normalizedPath = path.startsWith('/') ? path.substring(1) : path;
    return '$normalizedBase$normalizedPath';
  }

  Future<HttpClientRequest> _openRequest(String method, Uri uri) {
    switch (method) {
      case 'POST':
        return _httpClient.postUrl(uri);
      case 'PATCH':
        return _httpClient.patchUrl(uri);
      case 'DELETE':
        return _httpClient.deleteUrl(uri);
      default:
        return _httpClient.getUrl(uri);
    }
  }

  void _assertSecureBaseUri(Uri uri) {
    final isLocalhost = uri.host == 'localhost' || uri.host == '127.0.0.1' || uri.host == '10.0.2.2';
    if (uri.scheme != 'https' && !isLocalhost) {
      throw StateError('Buddies API requires HTTPS for non-local network communication.');
    }
  }
}

String sha256Hex(String value) => sha256.convert(utf8.encode(value)).toString();

class LoginResponse {
  const LoginResponse({
    required this.accessToken,
    required this.tokenType,
    required this.expiresIn,
  });

  final String accessToken;
  final String tokenType;
  final int expiresIn;

  factory LoginResponse.fromJson(Map<String, dynamic> json) {
    return LoginResponse(
      accessToken: json['accessToken'] as String,
      tokenType: json['tokenType'] as String,
      expiresIn: json['expiresIn'] as int,
    );
  }
}

class LobbySummary {
  const LobbySummary({
    required this.id,
    required this.restaurantName,
    required this.deliveryLocation,
    required this.orderStatus,
    required this.cartLocked,
  });

  final int id;
  final String restaurantName;
  final String deliveryLocation;
  final String orderStatus;
  final bool cartLocked;

  factory LobbySummary.fromJson(Map<String, dynamic> json) {
    return LobbySummary(
      id: json['id'] as int,
      restaurantName: json['restaurantName'] as String,
      deliveryLocation: json['deliveryLocation'] as String,
      orderStatus: json['orderStatus'] as String,
      cartLocked: json['cartLocked'] as bool,
    );
  }
}

class ApiException implements Exception {
  const ApiException(this.statusCode, this.body);

  final int statusCode;
  final String body;

  @override
  String toString() => 'ApiException($statusCode): $body';
}

File: admin-web/src/apiClient.ts
export type SystemOverview = {
  activeLobbyCount: number;
  cartLockedLobbyCount: number;
  activeUserCount: number;
  openReportCount: number;
  openSupportTicketCount: number;
  suspendedUserCount: number;
  recentLobbies: RecentLobby[];
};

export type ReportPage = {
  items: ReportSummary[];
  page: number;
  size: number;
  totalCount: number;
};

export type ReportSummary = {
  reportId: number;
  lobbyId: number;
  reporterUserId: number;
  reportedUserId: number;
  reason: string;
  status: string;
  createdAt: string;
};

export type ReportDetail = {
  reportId: number;
  lobbyId: number;
  reporter: UserReference;
  reportedUser: UserReference;
  reportedMessageId: number | null;
  reason: string;
  description: string | null;
  status: string;
  resolutionNote: string | null;
  createdAt: string;
};

export type UserReference = {
  id: number;
  name: string;
};

export type ChatArchive = {
  lobbyId: number;
  messages: ArchiveMessage[];
};

export type ArchiveMessage = {
  messageId: number;
  lobbyId: number;
  senderUserId: number | null;
  senderName: string | null;
  messageType: string;
  content: string | null;
  mediaUrl: string | null;
  reported: boolean;
  createdAt: string;
};

export type RecentLobby = {
  lobbyId: number;
  restaurantName: string;
  deliveryLocation: string;
  hostUserId: number;
  participantCount: number;
  currentTotal: number;
  orderStatus: string;
  cartLocked: boolean;
  createdAt: string;
};

export type CurrentUser = {
  id: number;
  email: string;
  name: string;
  role: string;
};

export type AdminUserPage = {
  items: AdminUserSummary[];
  page: number;
  size: number;
  totalCount: number;
};

export type AdminUserSummary = {
  id: number;
  email: string;
  name: string;
  role: string;
  status: string;
  trustScore: number;
  createdAt: string;
};

export type AdminUserDetail = AdminUserSummary & {
  reportedCount: number;
  reporterCount: number;
  closedLobbyCount: number;
  moderationActions: ModerationAction[];
};

export type ModerationAction = {
  id: number;
  actionType: string;
  reason: string;
  adminUserId: number;
  adminName: string;
  reportId: number | null;
  startsAt: string;
  endsAt: string | null;
  createdAt: string;
};

export type SupportTicketPage = {
  items: SupportTicketSummary[];
  page: number;
  size: number;
  totalCount: number;
};

export type SupportTicketSummary = {
  ticketId: number;
  userId: number;
  userName: string;
  lobbyId: number | null;
  category: string;
  title: string;
  status: string;
  createdAt: string;
  updatedAt: string;
};

export type SupportTicketDetail = {
  ticketId: number;
  user: UserReference;
  lobbyId: number | null;
  category: string;
  title: string;
  body: string;
  status: string;
  resolutionNote: string | null;
  resolvedByAdmin: UserReference | null;
  resolvedAt: string | null;
  createdAt: string;
  updatedAt: string;
};

type RequestOptions = {
  token?: string;
  query?: Record<string, string | number | boolean | undefined>;
};

type RequestBody = Record<string, unknown>;

export class ApiClient {
  constructor(private readonly baseUrl = '/api') {
    this.assertSecureBaseUrl(baseUrl);
  }

  async getSystemOverview(options: RequestOptions = {}): Promise<SystemOverview> {
    return this.request<SystemOverview>('/admin/system/overview', options);
  }

  async getReports(status = 'OPEN', page = 1, size = 20, options: RequestOptions = {}): Promise<ReportPage> {
    return this.request<ReportPage>('/admin/reports', {
      ...options,
      query: { status, page, size },
    });
  }

  async getReport(reportId: number, options: RequestOptions = {}): Promise<ReportDetail> {
    return this.request<ReportDetail>(`/admin/reports/${reportId}`, options);
  }

  async resolveReport(reportId: number, resolutionNote: string, options: RequestOptions = {}): Promise<MessageResponse> {
    return this.request<MessageResponse>(`/admin/reports/${reportId}/resolve`, {
      ...options,
      method: 'PATCH',
      body: { resolutionNote },
    });
  }

  async getSupportTickets(status = 'OPEN', page = 1, size = 20, options: RequestOptions = {}): Promise<SupportTicketPage> {
    return this.request<SupportTicketPage>('/admin/support-tickets', {
      ...options,
      query: { status, page, size },
    });
  }

  async getSupportTicket(ticketId: number, options: RequestOptions = {}): Promise<SupportTicketDetail> {
    return this.request<SupportTicketDetail>(`/admin/support-tickets/${ticketId}`, options);
  }

  async updateSupportTicket(ticketId: number, status: string, resolutionNote: string, options: RequestOptions = {}): Promise<MessageResponse> {
    return this.request<MessageResponse>(`/admin/support-tickets/${ticketId}`, {
      ...options,
      method: 'PATCH',
      body: { status, resolutionNote: resolutionNote || undefined },
    });
  }

  async getChatArchive(lobbyId: number, options: RequestOptions = {}): Promise<ChatArchive> {
    return this.request<ChatArchive>(`/admin/lobbies/${lobbyId}/chat-archive`, options);
  }

  async login(email: string, password: string): Promise<LoginResponse> {
    // Passwords are sent only over HTTPS/TLS and must be bcrypt-verified by the server.
    return this.request<LoginResponse>('/auth/login', {
      method: 'POST',
      body: { email, password },
    });
  }

  async getMe(options: RequestOptions = {}): Promise<CurrentUser> {
    return this.request<CurrentUser>('/auth/me', options);
  }

  async getUsers(status = 'ACTIVE', page = 1, size = 20, options: RequestOptions = {}): Promise<AdminUserPage> {
    return this.request<AdminUserPage>('/admin/users', {
      ...options,
      query: { status, page, size },
    });
  }

  async getUser(userId: number, options: RequestOptions = {}): Promise<AdminUserDetail> {
    return this.request<AdminUserDetail>(`/admin/users/${userId}`, options);
  }

  async moderateUser(
    userId: number,
    actionType: string,
    reason: string,
    endsAt: string | null,
    reportId: number | null,
    options: RequestOptions = {},
  ): Promise<MessageResponse> {
    return this.request<MessageResponse>(`/admin/users/${userId}/moderation-actions`, {
      ...options,
      method: 'POST',
      body: { actionType, reason, endsAt: endsAt ?? undefined, reportId: reportId ?? undefined },
    });
  }

  async verifySignup(email: string, otp: string): Promise<MessageResponse> {
    return this.request<MessageResponse>('/auth/signup/verify', {
      method: 'POST',
      body: { email, otp },
    });
  }

  async confirmPasswordReset(token: string, newPassword: string, newPasswordConfirm: string): Promise<MessageResponse> {
    return this.request<MessageResponse>('/auth/password-reset/confirm', {
      method: 'POST',
      body: {
        token: await sha256Hex(token),
        newPassword,
        newPasswordConfirm,
      },
    });
  }

  private async request<T>(
    path: string,
    options: RequestOptions & { method?: string; body?: RequestBody } = {},
  ): Promise<T> {
    const response = await fetch(this.buildUrl(path, options.query), {
      method: options.method ?? 'GET',
      headers: this.headers(options.token),
      body: options.body ? JSON.stringify(options.body) : undefined,
      cache: 'no-store',
      credentials: 'same-origin',
    });

    if (!response.ok) {
      throw new ApiError(response.status, await this.errorMessage(response));
    }

    return response.json() as Promise<T>;
  }

  private buildUrl(path: string, query?: RequestOptions['query']) {
    const url = new URL(`${this.baseUrl}${path}`, window.location.origin);
    Object.entries(query ?? {}).forEach(([key, value]) => {
      if (value !== undefined) {
        url.searchParams.set(key, String(value));
      }
    });
    return url;
  }

  private headers(token?: string) {
    const headers = new Headers({
      Accept: 'application/json',
      'Content-Type': 'application/json',
      'Cache-Control': 'no-store',
    });
    if (token) {
      headers.set('Authorization', `Bearer ${token}`);
    }
    return headers;
  }

  private async errorMessage(response: Response) {
    try {
      const body = await response.json();
      return body.message ?? body.error ?? response.statusText;
    } catch {
      return response.statusText;
    }
  }

  private assertSecureBaseUrl(baseUrl: string) {
    const url = new URL(baseUrl, window.location.origin);
    const isLocalhost = ['localhost', '127.0.0.1', '::1'].includes(url.hostname);
    if (url.protocol !== 'https:' && !isLocalhost) {
      throw new Error('Buddies API requires HTTPS for non-local network communication.');
    }
  }
}

export class ApiError extends Error {
  constructor(
    readonly status: number,
    message: string,
  ) {
    super(message);
  }
}

export type LoginResponse = {
  accessToken: string;
  tokenType: 'Bearer';
  expiresIn: number;
};

export type MessageResponse = {
  message: string;
};

export async function sha256Hex(value: string): Promise<string> {
  const bytes = new TextEncoder().encode(value);
  const digest = await crypto.subtle.digest('SHA-256', bytes);
  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('');
}
